### PR TITLE
[apex] Add property to allow apex test methods to contain underscores

### LINKED ISSUE #1573: [apex] allow test method names to contain underscores
Firstly, PMD is great - thank you!

I have a feature request for the Salesforce Apex language, regarding the warning "Method names should not contain underscores"

I would be great if this warning did not appear if a method is marked as 'testMethod' or '@istest'

Because using underscores in unit test method names is a common practice to make them readable.

Would this be possible?

### PR DESCRIPTION
**Pull request by tomdaly**
Add a 'skipTestMethodUnderscore' boolean property that allows an
annotated Apex test method (@isTest or testMethod) to contain underscores
Fixes #1573

Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

### CODE REVIEW

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml [hunk 1]
```diff
@@ -66,4 +66,16 @@ public class Foo {
...
     </test-code>
     
+    <test-code>
+        <description>#1573 method names should not contain underscores, but skip test methods</description>
+        <expected-problems>0</expected-problems>
+        <rule-property name="skipTestMethodUnderscores">true</rule-property>
+        <code><![CDATA[
+public class Foo {
+    @isTest
+	void test_barFoo() {}
```
Comment (jsotuyod) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml]: please. revise the indentation here

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml [hunk 2]
```diff
@@ -66,4 +66,16 @@ public class Foo {
         ]]></code>
     </test-code>
     
+    <test-code>
+        <description>#1573 method names should not contain underscores, but skip test methods</description>
+        <expected-problems>0</expected-problems>
+        <rule-property name="skipTestMethodUnderscores">true</rule-property>
```
Comment (jsotuyod) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml]: we are missing an equivalent test with this property set to false, and checking we get 1 violation reported

Comment (tomdaly) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml]: Thanks for the comments - should I add three more testcases to represent the different test annotations?
1 (current test case) - @isTest, property true, 0 expected problems
2 - @isTest, property false, 1 expected problem
3 - testMethod property true, 0 expected problems
4 - testMethod, property false, 1 expected problem

or would 1, 2 and 3 be enough to cover the bases?

Comment (jsotuyod) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/MethodNamingConventions.xml]: Having the 4 would be most complete, but 3 is probably good enough for coverage

**Comment by jsotuyod** [comment]
@tomdaly this is now merged and will be part of PMD 6.12.0. Thanks again for your work!

**Comment by joeythomaschaske** [comment]
Why has this been deprecated? https://pmd.github.io/latest/pmd_rules_apex_codestyle.html#methodnamingconventions

**Comment by catalandres** [comment]
@joeythomaschaske I believe it's because there is a more flexible way to achieve the same, using regex. Check [#1817](https://github.com/pmd/pmd/pull/1817).

**Comment by adangel** [comment]
Exactly, use the regex property `testPattern` to allow/disallow underscores in test method names. By default, underscore are not allowed. E.g. by using the following configuration, you can allow underscores:

```xml
<rule ref="category/apex/codestyle.xml/MethodNamingConventions">
    <properties>
        <property name="testPattern" value="[a-z][a-zA-Z0-9_]*" />
    </properties>
</rule>
```

**Pull request merged**